### DISCUSSION TITLE
Code compilation - Docker version

### DISCUSSION THREAD

**Discussion by spottier88** [hsakoh/switchbot-mqtt#2] [Q&A] [answered]
Hello, first of all, thank you for your work!
I was looking for a solution for switchbot without BLE for a while.

I want to use your project in docker (for use with another domotics solution, via MQTT) but I don't understand if I need to compile the project on linux before running it in docker or if I can compile it on windows, put the compiled project in my docker VM, and then enable the docker image in my VM.

Can you help me ?

I hope to be specific in my question. 

Thank you in advance

**Comment by hsakoh** [comment]
I'm not familiar with Docker or file handling, so I guess I confused you.

There are two steps: building the application and building the Docker image.
For the application, it can be built on both Linux and Windows as long as you have .NET 6 SDK.
Regarding the Docker image, it can also be built on both Linux and Windows.

The provided Dockerfile seems like it can be used almost as-is.
You can find the Dockerfile [here](https://github.com/hsakoh/switchbot-mqtt/blob/main/_build_on_haos/Dockerfile)

Here are the steps to build & execute on WSL2.
```
git clone https://github.com/hsakoh/switchbot-mqtt.git
cd switchbot-mqtt
cp ./_build_on_haos/Dockerfile ./
docker build ./ -t my-image --build-arg BUILD_FROM=homeassistant/amd64-base:latest --build-arg BUILD_ARCH=amd64 --progress=plain

mkdir ./data/
cp ./_docker-compose/options_Example.json ./data/options.json

#configure ./data/options.json 
# SwitchBot ApiKey,ApiSecret and WebhookService NgrokAuthToken ..etc..

docker run --name switchbot-mqtt -p=8099:8099/tcp -p=8098:8098/tcp -v ./data:/data my-image

#Open http://localhost:8099
```

Here are the instructions for pulling the pre-built image from the GitHub Container Registry & execute .
```
docker pull ghcr.io/hsakoh/switchbotmqtt-amd64:1.0.1

mkdir ./switchbot-mqtt-data
wget https://raw.githubusercontent.com/hsakoh/switchbot-mqtt/main/_docker-compose/options_Example.json -o ./switchbot-mqtt-data/options.json 

#configure ./switchbot-mqtt-data/options.json 
# SwitchBot ApiKey,ApiSecret and WebhookService NgrokAuthToken ..etc..

docker run --name switchbot-mqtt -p=8099:8099/tcp -p=8098:8098/tcp -v ./switchbot-mqtt-data/:/data my-image
```

**Comment by hsakoh** [comment, marked as answer]
I have updated the installation instructions (added as an advanced installation method).
https://github.com/hsakoh/switchbot-mqtt/wiki/HowToInstall#advanced-add-on-installation-method
